Introduce config validators

kokorin opened the pull request:
As discussed in #150

littleK0i commented:
In theory, it is nice to have `validators` as a separate entity.

There is a small problem. Parsers are currently doing some validation and having some dependencies on each other. For example, this place in code relies on `SchemaParser` being executed before `BusinessRoleParser` for pattern matching: https://github.com/littleK0i/SnowDDL/blob/master/snowddl/parser/business_role.py#L190

We might be avoid resolving patterns in parsers and do it in validator followed by resolver. But it also means doing pattern matching twice.

Need to think about it a bit and look for other potential references.

littleK0i commented:
On the other hand, we could potentially add one more generic `validate` method to parsers and run it in another loop.

Something like..

```
        # Blueprints
        for parser_cls in self.parse_sequence:
            parser = parser_cls(config, scanner)
            parser.load_blueprints()

        if config.errors:
            self.output_config_errors(config)
            exit(1)

        # Validate
        for parser_cls in self.parse_sequence:
            parser = parser_cls(config, scanner)
            parser.validate_blueprints()

        if config.errors:
            self.output_config_errors(config)
            exit(1)
```

Sequence will remain the same, but validation will happen after everything else was loaded. We may assume that full config is available at the time of validation.

kokorin commented:
I thought about project structure and can suggest the following changes:

1. Introduce Pydantic-based dataclass for each SnowDDL entity, for example `DatabaseConf` or `DatabaseYaml`.
It would allow to use Pydantic-based validation on YAML content (without  cross-references). JSON Schema definition will not be needed.
2. Introduce Validators to validate cross-references in confi dataclasses.
It makes sense to run Validators on dataclasses (not Blueprints) to validate config as early as possible.
3. Create Bluprints based on conf dataclasses
4. Refactor PermissionModel class, it's the cornerstone of DB/Schema.
As of now, PM is passive, the only significant method in it is `build_permission_model` which is used to support PM inheritance.
In several places there are `if`-clauses like `if database_permission_model.ruleset.create_database_owner_role:` which complicates introduction of another (potential) PM ruleset.
I suggest instead to make PM active: PM should decide which privileges to grant on which object. For that add 2 subclasses: `SchemaOnwerPM` and `DatabaseOwnerPM`, each one implementing behavior specific for each ruleset. As a plus it would simplify introduction of new ruleset in future.

I understand it's huge amount of work, but we can plan on splitting it into small tasks and of course I'm willing to help (19 years of development experience)

littleK0i commented:
Pydantic-based validation for YAML might be a good idea, especially for major version upgrade (0.x.x to 1.x.x). When I started SnowDDL, pydantic v2 did not exist and pydantic v1 was much worse. The original implementation was based on standard dataclasses and `jsonschema`. Eventually I've switched blueprints to pydantic v2. In theory, we might be able to get rid of `jsonschema` and use pydantic v2 for everything.

Json schema produces relaitvely good quality errors for nested data structures. I am not sure if pydantic can properly report nesting. I suspect, multiple models will be required to repesent nesting. Potentially it might be harder to find a problem in large files, e.g. `users.yaml` with hundreds of users. Need to investigate.

---

In fact, running validators on blueprints is better because [programmatic config](https://docs.snowddl.com/advanced/programmatic-config) is an important feature. It is not uncommon to generate blueprints for 50%+ objects in large accounts. YAML is not the only source of config, far from it.

We definitely want to validate both objects from YAML configs and auto-generated objects together.

---

Permission model is an interesting topic. Originally we had only one schema-based permission model. It works very well in controlled environments, and we never ran into any issues. When you fully control your data stack, it is amazing.

But it appears tools like Fivetran and Airbyte became very popular in Snowflake ecosystem. These tools are poorly designed on fundamental level, but it seems people do not know any better and keep using them.

The main problem with these tools is hard requirement to be able to create their own schemas. No schema parameters can be specified, no future grants created, nothing. They just want to run `CREATE SCHEMA` and fail if it is not allowed.

In order to accomodate that I've added "database_owner" permission model. But I currently have no plans to expand this feature and grant even more freedom. Snowflake has very complicated permission system, and it is super-easy to shoot yourself in the foot. Lack of freedom is intentional.

But I am open to change my mind if:
- You have a good use-case which is not covered by existing schema-centric or database-centric permission models;
- It cannot be mitigated by `global_roles` easily;
- It will not create other obscure problems;

kokorin commented:
> In fact, running validators on blueprints is better because [programmatic config](https://docs.snowddl.com/advanced/programmatic-config) is an important feature. It is not uncommon to generate blueprints for 50%+ objects in large accounts. YAML is not the only source of config, far from it.

I played a bit with programmatic config and added custom blueprints, custom resolvers, and registered them (added) them in `default_resolve_sequence`. In that case it's not possible to validate custom blueprints. But I agree, that validating Blueprints may be better.

Have you considered supporting plugin system over single "handle" method? It may have sense for use-cases when (as in our case) we have many SF accounts and want to configure them using the same extra custom configuration?

> In order to accomodate that I've added "database_owner" permission model. But I currently have no plans to expand this feature and grant even more freedom. Snowflake has very complicated permission system, and it is super-easy to shoot yourself in the foot. Lack of freedom is intentional.

My suggestion to have new permission model was hypothetical, I do not have any use-case for new PM. My suggestion was to let PM decide on DB/schema grants, it can help with de-coupling logic to calculate grants and will also simplify unit-testing.

littleK0i commented:
@kokorin , did some tests related to validators.

This idea works fine for some blueprints, but it conflicts with the way how business / technical roles are currently implemented. All role blueprints are currently very simple, and all the grant-building is happening directly inside parsers. For example: https://github.com/littleK0i/SnowDDL/blob/master/snowddl/parser/business_role.py#L110-L148

Maybe I should rework these roles before validators can be implemented. More specifically, move grant-building to resolvers from parsers.

Alternatively, we may simply add some more validation to existing parsers and ignore validation for blueprints generated in code.

---

About "plugins". Currently I see two extreme use-cases for custom coding in config:

1. "Need to change a little bit", so basic files with `handle()` should be fine.
2. "Need to change a lot, including new parsers / resolver, new object types, major adjustments to execution logic, etc." In this case it might be easier to clone & fork an entire SnowDDL.

The main advantage of Python is how easy it is to clone & change stuff. Not sure if we need to add extra systems on top of it when language gives so much freedom by design.

littleK0i commented:
@kokorin , please take a look at the latest commits, specifically at changelog and validators: https://github.com/littleK0i/SnowDDL/tree/master/snowddl/validator

I had to rework a lot of things related to parsing and config management, but now easy validation is possible both for YAML config and programmatic config.

Not yet on PyPi, planning to deploy it tomorrow or on Friday after documentation updates.
